### PR TITLE
feat: Binary tree node sum (#7020)

### PR DESCRIPTION
**Pull request by CaedenPH**
### Describe your change:
Implements #7020 
Algorithm to get the sum of all tree nodes

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by CaedenPH** [comment]
@cclauss pinging for review

File: data_structures/binary_tree/binary_tree_node_sum.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    >>> tree.right.left = Node(8)
+    >>> tree.right.right = Node(0)
+    >>> BinaryTreeNodeSum(tree).node_sum()
+    32
+    """
+
+    def __init__(self, tree: Node) -> None:
+        self.tree = tree
+
+    def depth_first_search(self, node: Node | None) -> int:
```
Comment (cclauss) [inline data_structures/binary_tree/binary_tree_node_sum.py:61]: from collections.abc import Iterator

Extra credit: `def __iter__(self) -> Iterator[int]`

Then you could replace `node_sum()` with `sum(binary_tree_node_sum)`.

**Review by cclauss — APPROVED** [review]
Let me know if you want to try the extra credit bit or not.

**Comment by CaedenPH** [comment]
@cclauss 
Is this what you meant?

**Comment by dhruvmanila** [comment]
> Is this what you meant?

No, that's not what he meant. He's asking you to implement the iteration protocol for a custom Python object. It comprises of two methods `__iter__` to which returns an iterator and `__next__` which returns the next object.

These methods would be made available to the `Node` class which then would allow us to do:
```python
sum(node.value for node in tree)
```

If you need a bit more understanding, I could help you out with an example.

A class for doing a sum seems redundant. Alternatively, you could just create a function instead.

**Comment by cclauss** [comment]
`list(binary_tree_node_sum)` would return `[10, 5, 12, -3, 8, 0]`

`sum(binary_tree_node_sum)` would return `32`

 I would just define `__iter__` and not define `__next__` but whatever works for you.  I would use `yield` more than return.  There are lots of examples in this repo that I linked to above.

**Comment by CaedenPH** [comment]
> `list(binary_tree_node_sum)` would return `[10, 5, 12, -3, 8, 0]`
> 
> 
> 
> `sum(binary_tree_node_sum)` would return `32`.
> 
> 
> 
> `max()`, `min()`, `mean()` would all just work out-of-the-box.
> 
> 
> 
>  I would just define `__iter__` and not define `__next__` but whatever works for you.  I would use `yield` more than `return`.  There are lots of examples in this repo that I linked to above.

Okay, I understand. However, perhaps all of this should be implemented in the basic binary tree or some other more general file instead of binary tree path sum

**Pull request merged**